lib/crawly/asset_downloader.ex

```diff
@@ -0,0 +1,44 @@
+defmodule Crawly.AssetDownloader do
+  require Logger
+
+  use GenServer
+
+  def start_link([]) do
+    Logger.debug("Starting asset downloader")
+
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  @impl true
+  def init(_args) do
+    {:ok, %{}}
+  end
+
+  def download_asset(directory, url) do
+    GenServer.cast(__MODULE__, {:download_asset, directory, url})
+  end
+
+  @impl true
+  def handle_cast({:download_asset, directory, url}, state) do
+    :ok = File.mkdir_p(directory)
+
+    case HTTPoison.get(url) do
+      {:ok, response} ->
+        full_path = Path.join(directory, file_name(url))
+        :ok = File.write(full_path, response.body)
+        Logger.debug("Downloaded image from URL #{url} into #{full_path}")
+      {:error, error} ->
+        Logger.error("Could not download image from URL #{url}: #{inspect(error)}")
+    end
+
+    {:noreply, state}
+  end
+
+  # Returns the string after the last slash, e.g.:
+  # https://amazon.com/products/picture.jpg => picture.jpg
+  # It is usually a good approximation of a "file name" on the web
+  defp file_name(url) do
+    Regex.run(~r/(?<=\/)[^\/]+$/, url)
```

Comment:
I would suggest storing images using guid as a file name. Otherwise we will get duplicates and overrides. It's better to extend item with a field called downloaded images
.

Reply:
I don't understand the last sentence, could you please explain?

Reply:
I suggest generating image filenames in the pipeline(`download_media.ex`) so the assets downloader will get two things
1. URL to download from
2. path to store

The item pipeline could look this way in the case:
```
def run(item, state, opts \\ []) do
.....
AssetDownloader.download_asset(filename, image_url)
{Map.put(item, :image_path, filename) , state}
end
```
In this case, the item structure will already have the field image_path which will point to the actual image.